README.md
CryptoFX

## Current State
Pre-Alpha v0.1.0

Open to any collaborators, help, advice, code review, issue, insult, critism, hugs or beer :)

## Overview

The main goal of CryptoFX is to allow create a trading online based and social platform.
It should be able to automatize most of

## Technology behind
* **NodeJS** (trying to use last NodeJS tech -> node.green) (https://nodejs.org)
* **Bluebird** - Optimized promise library (http://bluebirdjs.com) 
* **asyncawait** - Callback heaven for Node.js with async/await (https://github.com/yortus/asyncawait)
* **moment** - Parse, validate, manipulate, and display dates in javascript. (http://momentjs.com)
* **Seneca** - A microservices toolkit for Node.js. (http://senecajs.org)
* ~~**ws** - Lightning fast websocket~~ (https://github.com/websockets/ws)
* MongoDB/LokiJS (to be defined),actual : In-Mem db ? (https://www.mongodb.com / lokijs.org)

#### Quick access :

-  [Documentation](DOCUMENTATION.md)
-  [Changelog](CHANGELOG.md)
-  [Roadmap](ROADMAP.md)
-  [ToDos](ROADMAP.md#todos)
-  [NPM Package file](package.json)

## Install
```npm install```

## Testing 
```node tests/test.js```

## Start 
On windows :
 - Navigate to root ```cd CryptoFX```
 - Start CMD (from root) ```"bin/win/start_all.bat"
 It will then open one CMD per node(microservices), theses microservices being connected by seneca 

## MicroServices 

- FETCHER : Call API (see Markets) and send data to MIB
- MIB : Mainly used as a tick store | HTTP : 12121
- MAB : Mainly used to store analyzed tick (Indicators, MarketSignals)

To be born : 

- ~~CARL : Main interface that will operate user needs~~
- SmartAnalyst : 
    At first launch, Get data from MIB, analyze them, compute them in terms of tickframe and send it to MAB
    after that, periodically get misisng data from MIB and MAB.
- TradingOperator : Able to Buy or Sell
- WalletManager : Able to handle Wallet (virtual ?)
- CryptoStrategist : Understand strategy and signals.

## Critical TODO
    - Implement databse !


## Markets Exchanges handled : 

- Bittrex : Honestly, It could work... But there is an issues with caching that is not resolved yet. So consider it down. 
- Poloniex : Handled for Nth (configurated) top Market (30s between two ticks fetch)   
- Kraken : Upcoming Prio 2! 

## Main Features implemented

* Get data (fetch) from Bittrex API
* Execute fetching at the next plain hours (will be useful for data analysis)
* Populate tick in MIB Db.
* Fetch tick from MIB to MAB
* Create Candle (open, close, mid, bid, ask, last, low, high, BTCVol)
* Calculate Technical Indicators:
    - SMA(10), SMA(20), SMA(50), SMA(100), SMA(200)
    - True stength index (25,13)
    - EMA(27)
* Calculate MarketSignals : 
    - largeGoldenCross(200,50)
    - largeDeathCross(200,50)

## Upcoming Features / Schedule

* Be able to run it on a local PC first (server will come next, but well, It just need a .sh instead of a .bat).
* Be able to connect to one of the trading online platform (kraken).
* Be able to gives instructions, modify a parameter using command line
* Be able to do same than above using an admistration interface
* Be able to use it in a meteor app
* Create an API
* Store data in a Raw Database
* Handle data from raw DB and analyze them in another db (treated db)
* In-browser graph (tradingview/meta-trader style)
* VirtualWallet
* RealWallet
* Indicators
* WebBased interface

### API 
# MIB (Raw DB) : 

* Store a tick :  Commands : role: 'MIB', store: 'ticker' | Params :  pair: {String}, exchange: {String}, ask {Number}, bid {Number}, last {Number} | Response : response {String}, dbSize {String}
* Get a tick : Commands : role: 'MIB', store: 'lastPrice' | Params :  pair: {String}, exchange: {String} | Response : response {String}, lastPrice {Object}, lastPrice.marketName {string}, lastPrice.price {number}, lastPrice.time} : {String}


## Contact

For any help or questions you can contact me by twitter : @obusco.

## Licence

This work is licensed under the Mozilla Public License Version 2.0

| Permissions               	| Conditions                                    	| Limitations   	|
|---------------------------	|-----------------------------------------------	|---------------	|
| :green_heart: Commercial Use 	| :grey_exclamation: Disclose Source              	| :broken_heart: Hold Liable   	|
| :green_heart: Distribution  	| :grey_exclamation: License and Copyright Notice 	| :broken_heart: Trademark Use 	|
| :green_heart: Modification   	| :grey_exclamation: Same License                 	|               	|
| :green_heart:  Patent Use    	|                                                 	|               	|
| :green_heart: Private Use    	|                                                 	|               	|


CHANGELOG.md
## Changelog

# 0.1.0 - 29/09/2016
* [Indicators] EMA, TSI, SMA
* [MarketSignals] GoldenCross, DeathCross

# 0.0.1 - 24/09/2016
* [FETCHER] Ability to get tick from Bittrex and Poloniex
* [MAB] Ability to get data from MIB and transform tick into candle
* [MIB] Ability to store tick and responds all data or lastPrice of specific market


MAB/mab.js
var moment = require('moment');
Object.entries = require('object.entries');
var _ = require('lodash');
const Indicators = require('../helpers/indicators.js');

var config = {
    timeframes: [
        //If our tick are too big to fulfill less than 1mn
        //Eg : Each tick are separed by 1m, but we ask 20s, what do we do ?
        //For now, just comment them
        // {name: '20s', margin: 20000},
        {name: '40s', margin: 40000},
        {name: '1m', margin: 60000},
        {name: '2m', margin: 2 * 60 * 1000},
        {name: '5m', margin: 5 * 60 * 1000},
        {name: '15m', margin: 15 * 60 * 1000},
        {name: '30m', margin: 30 * 60 * 1000},
        {name: '1h', margin: 1 * 60 * 60 * 1000},
        {name: '2h', margin: 2 * 60 * 60 * 1000},
        {name: '3h', margin: 3 * 60 * 60 * 1000},
        {name: '4h', margin: 4 * 60 * 60 * 1000},
        {name: '6h', margin: 6 * 60 * 60 * 1000},
        {name: '8h', margin: 8 * 60 * 60 * 1000},
        {name: '1d', margin: 24 * 60 * 60 * 1000},
        {name: '1w', margin: 7 * 24 * 60 * 60 * 1000}
    ],
};

const MAB = {
    database: {
        initiated: false
    },
    createExchange: function (exchangeName) {
        MAB.database[exchangeName] = {
            exchangeName: exchangeName,
            markets: {}
        };
        cl("Created exchange " + exchangeName);
    },
    createMarket: function (exchangeName, marketName) {
        var market = {
            marketName: marketName,
            history: {}
        };
        for (var index in config.timeframes) {
            var tf = config.timeframes[index].name;
            market.history[tf] = {};
        }
        MAB.database[exchangeName].markets[marketName] = market;
        cl("Created market " + marketName);
    },
    createCandle: function (ticks, timeframe) {
        var candle = {
            timeFrame: timeframe.name,
            open: 0,
            close: 0,
            mid: 0,
            bid: 0,
            ask: 0,
            last: 0,
            low: 0,
            high: 0,
            volume: 0,
            BTCVol: 0,
            timestamp: 0,
            techIndic: {
                volatility: 0,
                change: 0,
                gain: 0,
                loss: 0,
                avg_gain: 0,
                avg_loss: 0,
                rs: 0,
                rsi: 0,
                true_range: 0,
                avg_true_range: 0,
                twelves_days_ema: 0,
                twenty_six_days_ema: 0,
                macd: 0,
                signal: 0,
                histogram: 0
            }
        };
        var processedTicks = 0;
        var startBTCVol = 0;
        var endBTCVol = 0;
        for (_tickIndex in ticks) {
            var _tick = ticks[_tickIndex];
            if (startBTCVol == 0) {
                startBTCVol = _tick.BTCVol;
            }
            if (processedTicks == 0) {
                candle.open = _tick.last;
                candle.low = parseFloat(_tick.bid);
            }
            candle.mid += parseFloat(parseFloat(_tick.bid) + parseFloat(_tick.ask - _tick.bid) / 2);
            candle.bid += parseFloat(_tick.bid);
            candle.ask += parseFloat(_tick.ask);
            candle.last = parseFloat(_tick.last);
            
            if (candle.high < _tick.ask) {
                candle.high = _tick.ask
            }
            if (candle.high < _tick.last) {
                candle.high = _tick.last
            }
            if (candle.low > _tick.low) {
                candle.low = _tick.low
            }
            if (candle.low > _tick.last) {
                candle.low = _tick.last
            }
            if (_tick.hasOwnProperty('BTCVol')) {
                
                endBTCVol = ((startBTCVol - _tick.BTCVol) > endBTCVol) ? startBTCVol - _tick.BTCVol : endBTCVol;
                candle.BTCVol = endBTCVol;
            }
            if (!_tick.hasOwnProperty('volume')) {
                candle.volume = -1;
            }
            ;
            candle.timestamp = _tick.timestamp;
            processedTicks++;
        }
        candle.mid = candle.mid / processedTicks;
        candle.bid = candle.bid / processedTicks;
        candle.ask = candle.ask / processedTicks;
        candle.formatedTime = moment(candle.timestamp).format('YYYY-MM-DD HH:mm:ss');
        candle.close = _tick.last;
        
        return candle;
        
    },
    initDatabaseFromList: function (list) {
        if (!MAB.database.initiated) {
            for (var exchange in list) {
                MAB.createExchange(exchange)
                for (var market in list[exchange].markets) {
                    MAB.createMarket(exchange, market);
                }
            }
            MAB.database.initiated = true;
            MAB.database.lastSync = moment().valueOf();
        }
        for (var exchange in list) {
            for (var market in list[exchange].markets) {
                // for(var indexTick in list[exchange].markets[market].history){
                //     var tick = list[exchange].markets[market].history[indexTick];
                //     cl(tick);
                // }
                MAB.populateDatabaseFromTick(exchange, market, list[exchange].markets[market].history);
            }
        }
        return true;
    },
    populateDatabaseFromTick: function (exchange, market, tickHistory) {
        var addCandle = function (exchange, market, timeframe, candle) {
            MAB.database[exchange].markets[market].history[timeframe.name][candle.timestamp] = candle;
            // MAB.database[exchange].markets[market].history[timeframe.name].push(candle);
        };
        for (var index in config.timeframes) {
            var tf = config.timeframes[index].name;
            //If our tick are too big to fulfill less than 1mn
            //Eg : Each tick are separed by 1m, but we ask 20s, what do we do ?
            // if(tf=="1m"){
            var timestampCount = 0;
            var nb = 0;
            var lastTimestamp = 0;
            var tickAccumulator = [];
            
            for (var _tickIndex in tickHistory) {
                var _tick = tickHistory[_tickIndex];
                if (lastTimestamp == 0 && _tickIndex == 0) {
                    lastTimestamp = _tick.timestamp;
                } else {
                    timestampCount = timestampCount + (_tick.timestamp - lastTimestamp);
                    nb++;
                    lastTimestamp = _tick.timestamp;
                    tickAccumulator.push(_tick);
                    if (timestampCount >= (config.timeframes[index].margin) * 0.05) {
                        var relativeMarginTime = timestampCount / config.timeframes[index].margin;
                        if (relativeMarginTime >= 0.95 && relativeMarginTime <= 1.1) {
                            var candle = MAB.createCandle(tickAccumulator, config.timeframes[index]);
                            addCandle(exchange, market, config.timeframes[index], candle);
                            
                            tickAccumulator = [], timestampCount = 0;
                            nb = 0;
                        }
                    }
                }
            }
            // var tickReadyToBeCandelized = {};
            // MAB.createCandle(tickToCandelize,"1m")         
            // return;
            // }
        }
        // cl(tickHistory);
    },
    performCalculation: function () {
        
        var prepareValuesForCalculation = function (period, list, actualIndexInList) {
            
            var listArr = Object.entries(list);
            var maxIndex = listArr.length - 1;
            var tickArr = [];
            var values = [];
            if(actualIndexInList){
                if((actualIndexInList-period)>=0){
                    tickArr = Object.entries(list).slice(actualIndexInList-period, actualIndexInList);
                }else{
                    tickArr = Object.entries(list).slice(0, actualIndexInList);
                }
            }else{
                tickArr = Object.entries(list).slice(maxIndex - (period), maxIndex);
            }
            tickArr.filter(function (tick) {
                values.push(tick[1].close);
            });
            return values;
            
        };
        
        for (var exchangeName in MAB.database) {
            if (exchangeName == "initiated" || exchangeName == "lastSync") {
                
            } else {
                var exchange = MAB.database[exchangeName];
                for (var marketName in exchange.markets) {
                    var market = exchange.markets[marketName];
                    for (var i = 0; i < Object.keys(market.history).length; i++) {
                        
                        var timeframeName = Object.keys(market.history)[i];
                        var timeframe = market.history[timeframeName]
                        
                        cl("New ", marketName, "Timeframe", timeframeName, Object.keys(timeframe).length);
                        
                        var lastTick = null, tick = null;
                        
                        for (var tickIndex = 0; tickIndex < Object.keys(timeframe).length; tickIndex++) {
                            var tickList = timeframe;
                            var tickName = Object.keys(timeframe)[tickIndex];
                            tick = tickList[tickName];
                            
                            var lastNtickName = function (n) {
                                return Object.keys(timeframe)[tickIndex - n];
                            }
                            var lastName = lastNtickName(1) || tickName;
                            var lastTick = tickList[lastName];
                            
                            var calculateTechIndic = function () {
                                tick.techIndic.change = tick.last - lastTick.last;
    
                                tick.techIndic.sma = {};
                                tick.techIndic.sma[10] = Indicators.SMA.calculate({
                                    period: 10,
                                    values: prepareValuesForCalculation(10, tickList, tickIndex)
                                });
                                tick.techIndic.sma[20] = Indicators.SMA.calculate({
                                    period: 20,
                                    values: prepareValuesForCalculation(20, tickList, tickIndex)
                                });
                                tick.techIndic.sma[50] = Indicators.SMA.calculate({
                                    period: 50,
                                    values: prepareValuesForCalculation(50, tickList, tickIndex)
                                });
                                tick.techIndic.sma[100] = Indicators.SMA.calculate({
                                    period: 100,
                                    values: prepareValuesForCalculation(100, tickList, tickIndex)
                                });
                                tick.techIndic.sma[200] = Indicators.SMA.calculate({
                                    period: 200,
                                    values: prepareValuesForCalculation(200, tickList, tickIndex)
                                });
                                tick.techIndic.sma['bullBurst'] = false;
                                tick.techIndic.sma['bearBurst'] = false;
    
                                if (tickIndex == 200 + 1) {
                                    if (tick.techIndic.sma[50] > tick.techIndic.sma[200]) {
                                        if (tickList[lastNtickName(1)].techIndic.sma[50] == tickList[lastNtickName(1)].techIndic.sma[200]) {
                                            if (tickList[lastNtickName(2)].techIndic.sma[50] < tickList[lastNtickName(2)].techIndic.sma[200]) {
                                                tick.techIndic.sma['bullBurst'] = true;
                                            }
                                        }
                                        if (tickList[lastNtickName(1)].techIndic.sma[50] < tickList[lastNtickName(1)].techIndic.sma[200]) {
                                            tick.techIndic.sma['bullBurst'] = true;
                                        }
                                    }
                                    if (tick.techIndic.sma[50] < tick.techIndic.sma[200]) {
                                        if (tickList[lastNtickName(1)].techIndic.sma[50] == tickList[lastNtickName(1)].techIndic.sma[200]) {
                                            if (tickList[lastNtickName(2)].techIndic.sma[50] > tickList[lastNtickName(2)].techIndic.sma[200]) {
                                                tick.techIndic.sma['bearBurst'] = true;
                                            }
                                        }
                                        if (tickList[lastNtickName(1)].techIndic.sma[50] > tickList[lastNtickName(1)].techIndic.sma[200]) {
                                            tick.techIndic.sma['bullBurst'] = true;
                                        }
            
                                    }
        
                                }
                                if(tick.techIndic.sma['bullBurst'] || tick.techIndic.sma['bearBurst']){
                                    cl("OUIIII!!");
                                    cl(tick, lastTick);
                                }
    
    
                                tick.techIndic.tsi = Indicators.TSI.calculate({
                                    longPeriod: 25,
                                    shortPeriod:13,
                                    values: prepareValuesForCalculation(50, tickList, tickIndex)
                                });
    
                                tick.techIndic.ema = {};
                                tick.techIndic.ema[27] = Indicators.EMA.calculate({
                                    period: 27,
                                    values: prepareValuesForCalculation(28, tickList, tickIndex)
                                });
                            };
                            calculateMarketSignals = function () {
    
                                tick.marketSignals = {};
                                tick.marketSignals.largeGoldenCross = (tick.techIndic.sma['bullBurst']);
                                tick.marketSignals.largeDeathCross = (tick.techIndic.sma['bearBurst']);
    
                            };
                            calculateTechIndic();
                            calculateMarketSignals();                                        
                            
                            // var calculIndicatorWithSpecificLastTicksLength = function(maxLastTicksLength, actualLastTickLength,tickIndex, indicator){
                            //     var handledIndicator = ['sma','priceChange'];
                            //     if(maxLastTicksLength!='undefined'){
                            //         if(handledIndicator.indexOf(indicator)>-1){
                            //             if(lastTicksLength>maxLastTicksLength && tickIndex>maxLastTicksLength){
                            //                 switch (indicator){
                            //                     case "sma":
                            //                         var sma=0;
                            //                         for(var k = 1;k<=maxLastTicksLength; k++){
                            //                             sma+=parseFloat(timeframe[Object.keys(timeframe)[tickIndex-k]].last);
                            //                         }
                            //                         sma = sma/maxLastTicksLength;
                            //                         return sma;
                            //                         break;
                            //                     case "priceChange":
                            //                         var pc = (tick.last-lastTick.last);
                            //                         return pc;
                            //                     case "doubleSmoothedAbsolutePC":
                            //                         var dsaPC = 1;
                            //                         return dsaPC;
                            //                         break;
                            //                     case "doubleSmoothedPC":
                            //                         var dsPC = 1;
                            //                         return dsPC;
                            //                         break;
                            //                     case "tsi":
                            //                         return tsi = 100 * (tick.techIndic.doubleSmoothedPC/tick.techIndic.doubleSmoothedAbsolutePC);
                            //                         return tsi;
                            //                         break;
                            //                     default:
                            //                         return null;
                            //                     break;
                            //                 }
                            //             }return null;
                            //         }return null;
                            //     }
                            //     return null;
                            // };
                            
                            
                            // var tickArr = Object.entries(ticks).slice(0, period+1);
                            // cl("Tick arr is ", tickArr.length);
                            // var values = [];
                            // var x = tickArr.filter(function(tick){
                            //     values.push(tick[1].close);
                            //     return values;
                            // });
                            // cl(x);
                            // tick.techIndic.ema[27]= Indicators.EMA.calculate({period:27, values:values});
                            //
                            // tick.techIndic.sma[10] = calculIndicatorWithSpecificLastTicksLength(10,lastTicksLength,tickIndex, 'sma');
                            // tick.techIndic.sma[20] = calculIndicatorWithSpecificLastTicksLength(20,lastTicksLength,tickIndex, 'sma');
                            // tick.techIndic.sma[50] = calculIndicatorWithSpecificLastTicksLength(50,lastTicksLength,tickIndex, 'sma');
                            // tick.techIndic.sma[100] = calculIndicatorWithSpecificLastTicksLength(100,lastTicksLength,tickIndex, 'sma');
                            // tick.techIndic.sma[200] = calculIndicatorWithSpecificLastTicksLength(200,lastTicksLength,tickIndex, 'sma');
                            //
                            // if(lastTicksLength>200){
                            //     tick.techIndic.sma["200_50"]=(tick.techIndic.sma[50]/tick.techIndic.sma[200]);
                            //     tick.techIndic.sma["bullBurst"]=(tick.techIndic.sma[50]/sma>1)? -1: 1;
                            // }
                            // tick.techIndic.priceChange = calculIndicatorWithSpecificLastTicksLength(2, lastTicksLength, tickIndex, 'priceChange');
                            //
                            // tick.techIndic.doubleSmoothedPC = calculIndicatorWithSpecificLastTicksLength(25, lastTicksLength, tickIndex, 'doubleSmoothedPC');
                            // tick.techIndic.doubleSmoothedAbsolutePC = calculIndicatorWithSpecificLastTicksLength(25, lastTicksLength, tickIndex, 'doubleSmoothedAbsolutePC');
                            // tick.techIndic.tsi = calculIndicatorWithSpecificLastTicksLength(25, lastTicksLength, tickIndex, 'tsi');
                            //
                            // tick.marketSignals = {
                            //     largeDeathCross:false,
                            //     largeGoldenCross:false
                            // };
                            
                            //FIXME : This is currently bugged.
                            //GoldenCross is triggered but when DeathCross should, Golden is trigered again.
                            // if(lastTick.techIndic.sma["bullBurst"]!=tick.techIndic.sma["bullBurst"]){
                            //     if(tick.techIndic.sma["bullBurst"]==-1){
                            //         tick.marketSignals.largeGoldenCross = true;
                            //     }
                            //     if(tick.techIndic.sma["bullBurst"]==1){
                            //         tick.marketSignals.largeDeathCross = true;
                            //
                            //     }
                            // }
                            //                          
                            // tick.techIndic.change= tick.last - lastTick.last;
                        }
                    }
                }
            }
        }
        
        // var h = MAB.database['poloniex'].markets['BTC_XMR'].history['20s'];
        // for(var i = 200; i<Object.keys(h).length; i++){
        //     cl(h[Object.keys(h)[i]]);
        // }
        var display = MAB.database['poloniex'].markets['BTC_XMR'].history['40s'];
        var last = display[Object.keys(display)[Object.keys(display).length - 1]];
        var lastN1 = display[Object.keys(display)[Object.keys(display).length - 2]];
        var lastN2 = display[Object.keys(display)[Object.keys(display).length - 3]];
        cl(last);
        // cl(lastN1);
        // cl(lastN2);
    }
};
module.exports = MAB;